samples/matter/common/src/diagnostic/diagnostic_logs_end_user.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "diagnostic_logs_end_user.h"
+
+#include <lib/support/CodeUtils.h>
+#include <system/SystemError.h>
+#include <zephyr/retention/retention.h>
+
+using namespace chip;
+
+namespace
+{
+const struct device *kEndUserDataMem = DEVICE_DT_GET(DT_NODELABEL(end_user_logs_retention));
+} /* namespace */
+
+CHIP_ERROR DiagnosticLogsEndUser::Init()
+{
+	if (mIsInitialized) {
+		return CHIP_NO_ERROR;
+	}
+
+	mCapacity = retention_size(kEndUserDataMem);
+	VerifyOrReturnError(mCapacity > kHeaderSize, CHIP_ERROR_INTERNAL);
+
+	int ret = retention_is_valid(kEndUserDataMem);
+
+	/* If data is invalid just clean the header data, as we are not sure which part is corrupted. */
+	if (ret != 1) {
+		ret = retention_write(kEndUserDataMem, 0, reinterpret_cast<uint8_t *>(&mCurrentSize),
+				      sizeof(mCurrentSize));
+		VerifyOrReturnError(ret == 0, System::MapErrorZephyr(ret));
+		ret = retention_write(kEndUserDataMem, sizeof(mCurrentSize), reinterpret_cast<uint8_t *>(&mDataBegin),
+				      sizeof(mDataBegin));
+		VerifyOrReturnError(ret == 0, System::MapErrorZephyr(ret));
+	}
+
+	ret = retention_is_valid(kEndUserDataMem);
```

Comment:
Right, I have to re-think this. I considered the case that the first `retention_is_valid` returned false, because the memory is empty, so we write the data header and try to check if this is valid again to make sure the control data was saved correctly. I missed that in case of success we do it twice without any reason.